### PR TITLE
pwalk, pwalkdir: fix walk vs remove race

### PR DESCRIPTION
**Pull request by kolyshkin**
In some cases, when file walk is racing with removal, the ENOENT is
received by filepath.Walk[Dir], rather than by the callback.

Do ignore such errors, except for the root directory (Walk argument).

Modify the doc accordingly, and add a couple of test cases.

The "Race" test case, when testing the code before the fix, fails 100%
of times for pwalk and 90% for pwalkdir. Alas, I was unable to make it
fail 100% of the times without significantly increasing the test
complexity.
    
Fixes: #199

### CODE REVIEW

**Comment by rhatdan** [comment]
LGTM
This is one where the tests probably took 10 times as much time to write as the fix.
@thaJeztah PTAL

**Review by rhatdan — APPROVED** [review]

**Review by thaJeztah — DISMISSED** [review]
overall LGTM (nice!), but found 2 redundant defers, and left a suggestion.

don't think they're necessarily show-stoppers, but I'll let @kolyshkin have a look before merging 👍

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

@rhatdan PTAL

File: pkg/pwalkdir/pwalkdir_test.go
Comment (kolyshkin) [reply pkg/pwalkdir/pwalkdir_test.go:76]: AFAIK Go is fine having a function and a variable of the same name (as long as the context is clear).

But, we have `max` and `min` functions in `go-selinux/selinux_linux.go` -- those do need to be conditionally removed. 

Update: I tried to do that (by moving min/max to a separate file only used for go < 1.21), and go complains:

```
[kir@kir-rhat go-selinux]$ go1.21.0 test -v .
# github.com/opencontainers/selinux/go-selinux [github.com/opencontainers/selinux/go-selinux.test]
./selinux_linux.go:664:22: max requires go1.21 or later (-lang was set to go1.19; check go.mod)
./selinux_linux.go:667:23: min requires go1.21 or later (-lang was set to go1.19; check go.mod)
FAIL	github.com/opencontainers/selinux/go-selinux [build failed]
```

I do not want to introduce any extras for this, so let's just add a TODO to remove these once go < 1.21 is no longer supported. Added in #206

File: pkg/pwalk/pwalk_test.go
Comment (kolyshkin) [reply pkg/pwalk/pwalk_test.go]: done

File: pkg/pwalk/pwalk_test.go
Comment (kolyshkin) [reply pkg/pwalk/pwalk_test.go:73]: Again, Go is fine having a function and a variable of the same name (as long as the context is clear, which it is in this case).

File: pkg/pwalkdir/pwalkdir_test.go
Comment (thaJeztah) [reply pkg/pwalkdir/pwalkdir_test.go:76]: My thinking was to just rename them. I don't think there's an urgency to switch over to the ones in stdlib (at least I don't expect them to magically perform massively better).

But renaming them prevents linters (and my IDE 😅) from flagging them as shadowing / colliding with a builtin.

Comment (thaJeztah) [reply pkg/pwalkdir/pwalkdir_test.go:76]: Oh! w.r.t. the warning; I'm guessing that Go will use the language version from your go.mod.

Go is no longer the version you're running, but downgrades (or upgrades) the language based on what's in your go.mod (which can include "download and install a different version of go)

**Pull request merged**
